File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsDropdown.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma set 13476:22316 — mirrors web/Dropdown.tsx + lib/ilds_dropdown.dart.

enum class IldsDropdownSize { Large, Medium }
enum class IldsDropdownRequiredIndicator { Text, Asterisk }

data class IldsDropdownOption(
    val label: String,
    val value: String,
    val isDisabled: Boolean = false,
)

@Composable
fun IldsDropdown(
    modifier: Modifier = Modifier,
    label: String? = null,
    placeholder: String = "Select option",
    options: List<IldsDropdownOption> = emptyList(),
    selectedValue: String? = null,
    onValueChange: (String) -> Unit = {},
    size: IldsDropdownSize = IldsDropdownSize.Large,
    isRequired: Boolean = false,
    requiredIndicator: IldsDropdownRequiredIndicator = IldsDropdownRequiredIndicator.Text,
    isDisabled: Boolean = false,
    isLoading: Boolean = false,
    isNegative: Boolean = false,
    errorText: String? = null,
    helperText: String? = null,
    prefixIcon: (@Composable () -> Unit)? = null,
    isOpen: Boolean? = null,
    onToggle: (() -> Unit)? = null,
    menuSectionLabel: String? = null,
    showMenuFooter: Boolean = true,
    menuSecondaryLabel: String = "Secondary button",
    menuPrimaryLabel: String = "Primary button",
    onMenuSecondary: (() -> Unit)? = null,
    onMenuPrimary: (() -> Unit)? = null,
) {
    var internalOpen by remember { mutableStateOf(false) }
    val expanded = isOpen ?: internalOpen
    val selectedOption = options.find { it.value == selectedValue }
    val hasError = (isNegative || errorText != null) && !isDisabled
    val helperContent = if (hasError) errorText ?: helperText else helperText

    val triggerHeight = if (size == IldsDropdownSize.Large) 48.dp else 40.dp
    val fontSize = if (size == IldsDropdownSize.Large) IldsTokens.fontSize14 else IldsTokens.fontSize12

    val borderColor = when {
        isDisabled -> IldsTokens.neutralCoolgray300
        isLoading -> IldsTokens.neutralCoolgray500
        hasError -> IldsTokens.errorRed600
        expanded -> IldsTokens.primaryOrange500
        else -> IldsTokens.neutralCoolgray500
    }
    val background = when {
        isDisabled -> IldsTokens.neutralCoolgray200
        else -> IldsTokens.globalWhite000
    }
    val displayColor = if (selectedOption != null) {
        IldsTokens.neutralCoolgray900
    } else {
        IldsTokens.neutralCoolgray500
    }

    fun toggle() {
        if (isDisabled || isLoading) return
        onToggle?.invoke()
        if (isOpen == null) internalOpen = !internalOpen
    }

    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(IldsTokens.sp4),
    ) {
        if (label != null) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp4),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = label,
                    style = TextStyle(
                        fontSize = IldsTokens.fontSize16,
                        lineHeight = 20.sp,
                        fontWeight = IldsTokens.fontWeightBold,
                    ),
                    color = if (isDisabled) IldsTokens.neutralCoolgray300 else IldsTokens.neutralCoolgray900,
                )
                if (isRequired) {
                    Text(
                        text = if (requiredIndicator == IldsDropdownRequiredIndicator.Asterisk) "*" else "(required)",
                        style = TextStyle(
                            fontSize = if (requiredIndicator == IldsDropdownRequiredIndicator.Asterisk) {
                                IldsTokens.fontSize12
                            } else {
                                10.sp
                            },
                            fontWeight = if (requiredIndicator == IldsDropdownRequiredIndicator.Asterisk) {
                                IldsTokens.fontWeightBold
                            } else {
                                IldsTokens.fontWeightRegular
                            },
                        ),
                        color = if (requiredIndicator == IldsDropdownRequiredIndicator.Asterisk) {
                            IldsTokens.errorRed700
                        } else {
                            IldsTokens.neutralCoolgray800
                        },
                    )
                }
            }
        }

        Box {
            Surface(
                onClick = { toggle() },
                modifier = Modifier
                    .fillMaxWidth()
                    .semantics { contentDescription = label ?: placeholder },
                enabled = !isDisabled && !isLoading,
                shape = RoundedCornerShape(
                    topStart = IldsTokens.radiusMedium,
                    topEnd = IldsTokens.radiusMedium,
                    bottomStart = if (expanded) IldsTokens.radiusNull else IldsTokens.radiusMedium,
                    bottomEnd = if (expanded) IldsTokens.radiusNull else IldsTokens.radiusMedium,
                ),
                color = background,
                border = BorderStroke(if (expanded) 2.dp else 1.dp, borderColor),
            ) {
                Row(
                    modifier = Modifier
                        .defaultMinSize(minHeight = triggerHeight)
                        .padding(horizontal = IldsTokens.sp12),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp8),
                ) {
                    prefixIcon?.invoke()
                    Text(
                        text = selectedOption?.label ?: placeholder,
                        modifier = Modifier.weight(1f),
                        style = TextStyle(
                            fontSize = fontSize,
                            lineHeight = 18.sp,
                            fontWeight = IldsTokens.fontWeightRegular,
                        ),
                        color = displayColor,
                        maxLines = 1,
                    )
                    if (hasError && !isLoading) {
                        Icon(
                            imageVector = Icons.Filled.Warning,
                            contentDescription = "Error",
                            tint = IldsTokens.errorRed600,
                            modifier = Modifier.size(20.dp),
                        )
                    }
                    if (isLoading) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(20.dp),
                            color = IldsTokens.primaryOrange500,
                            strokeWidth = 2.dp,
                        )
                    } else {
                        Icon(
                            imageVector = Icons.Filled.KeyboardArrowDown,
                            contentDescription = if (expanded) "Collapse menu" else "Expand menu",
                            tint = if (expanded) IldsTokens.primaryOrange500 else IldsTokens.neutralCoolgray500,
                            modifier = Modifier
                                .size(20.dp)
                                .rotate(if (expanded) 180f else 0f),
                        )
                    }
                }
            }

            if (expanded && options.isNotEmpty()) {
                IldsDropdownMenu(
                    sectionLabel = menuSectionLabel,
                    options = options,
                    selectedValue = selectedValue,
                    showFooter = showMenuFooter,
                    secondaryLabel = menuSecondaryLabel,
                    primaryLabel = menuPrimaryLabel,
                    onSelect = { value ->
                        onValueChange(value)
                        if (isOpen == null) internalOpen = false
                    },
                    onSecondary = onMenuSecondary,
                    onPrimary = onMenuPrimary,
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }

        if (!isDisabled && helperContent != null) {
            Text(
                text = helperContent,
                style = TextStyle(
                    fontSize = IldsTokens.fontSize12,
                    lineHeight = 16.sp,
                    fontWeight = IldsTokens.fontWeightRegular,
                ),
                color = if (hasError) IldsTokens.errorRed600 else IldsTokens.neutralCoolgray700,
            )
        }
    }
}

File: ios/Sources/ILDSDesignSystem/IldsDropdown.swift
import ILDSTokens
import SwiftUI

// Mirrors lib/ilds_dropdown.dart + web Dropdown.tsx.

public enum IldsDropdownSize: Sendable {
    case large, medium
}

public struct IldsDropdownOption: Identifiable, Sendable {
    public let id: String
    public let label: String
    public let value: String
    public let isDisabled: Bool

    public init(label: String, value: String, isDisabled: Bool = false) {
        self.id = value
        self.label = label
        self.value = value
        self.isDisabled = isDisabled
    }
}

public struct IldsDropdown: View {
    private let label: String
    private let placeholder: String
    private let options: [IldsDropdownOption]
    @Binding private var selectedValue: String?
    private let isEnabled: Bool
    private let isLoading: Bool
    private let errorText: String?
    private let helperText: String?
    private let size: IldsDropdownSize

    @State private var isOpen = false

    public init(
        label: String,
        placeholder: String,
        options: [IldsDropdownOption],
        selectedValue: Binding<String?>,
        isEnabled: Bool = true,
        isLoading: Bool = false,
        errorText: String? = nil,
        helperText: String? = nil,
        size: IldsDropdownSize = .large
    ) {
        self.label = label
        self.placeholder = placeholder
        self.options = options
        _selectedValue = selectedValue
        self.isEnabled = isEnabled
        self.isLoading = isLoading
        self.errorText = errorText
        self.helperText = helperText
        self.size = size
    }

    private var selectedOption: IldsDropdownOption? {
        guard let selectedValue else { return nil }
        return options.first { $0.value == selectedValue }
    }

    public var body: some View {
        let hasError = errorText != nil
        let colors = IldsDropdownColors.resolve(
            isOpen: isOpen,
            hasError: hasError,
            isEnabled: isEnabled,
            hasSelection: selectedOption != nil
        )
        let metrics = IldsDropdownMetrics(size: size)
        let bottomText = errorText ?? helperText

        VStack(alignment: .leading, spacing: ILDSTokens.sp4) {
            Text(label)
                .font(.system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightMedium))
                .foregroundStyle(ILDSTokens.neutralCoolgray500)

            VStack(spacing: 0) {
                Button {
                    guard isEnabled, !isLoading else { return }
                    isOpen.toggle()
                } label: {
                    HStack {
                        Text(selectedOption?.label ?? placeholder)
                            .font(.system(size: metrics.fontSize, weight: ILDSTokens.fontWeightRegular))
                            .foregroundStyle(colors.text)
                            .lineLimit(1)
                        Spacer()
                        if isLoading {
                            ProgressView()
                                .progressViewStyle(.circular)
                                .tint(ILDSTokens.primaryOrange500)
                                .frame(width: ILDSTokens.sp20, height: ILDSTokens.sp20)
                        } else {
                            Image(systemName: "chevron.down")
                                .font(.system(size: ILDSTokens.sp20))
                                .foregroundStyle(colors.icon)
                                .rotationEffect(.degrees(isOpen ? 180 : 0))
                                .animation(.easeInOut(duration: 0.15), value: isOpen)
                        }
                    }
                    .frame(height: metrics.triggerHeight)
                    .padding(.horizontal, ILDSTokens.sp16)
                    .background(colors.background)
                    .overlay {
                        IldsDropdownTriggerBorder(isOpen: isOpen, color: colors.border, width: colors.borderWidth)
                    }
                }
                .buttonStyle(.plain)
                .disabled(!isEnabled || isLoading)

                if isOpen {
                    IldsDropdownMenu(
                        options: options,
                        selectedValue: selectedValue,
                        onSelect: { value in
                            selectedValue = value
                            isOpen = false
                        }
                    )
                }
            }

            if let bottomText {
                Text(bottomText)
                    .font(.system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightRegular))
                    .foregroundStyle(hasError ? ILDSTokens.errorRed600 : ILDSTokens.neutralCoolgray400)
            }
        }
        .accessibilityLabel(label)
        .accessibilityValue(selectedOption?.label ?? placeholder)
    }
}

private struct IldsDropdownTriggerBorder: View {
    let isOpen: Bool
    let color: Color
    let width: CGFloat

    var body: some View {
        if isOpen {
            UnevenRoundedRectangle(
                topLeadingRadius: ILDSTokens.radiusMedium,
                topTrailingRadius: ILDSTokens.radiusMedium
            )
            .strokeBorder(color, lineWidth: width)
        } else {
            RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium)
                .strokeBorder(color, lineWidth: width)
        }
    }
}

private struct IldsDropdownMetrics {
    let triggerHeight: CGFloat
    let fontSize: CGFloat

    init(size: IldsDropdownSize) {
        switch size {
        case .large:
            triggerHeight = ILDSTokens.sp48
            fontSize = ILDSTokens.fontSize14
        case .medium:
            triggerHeight = ILDSTokens.sp40
            fontSize = ILDSTokens.fontSize12
        }
    }
}

private struct IldsDropdownColors {
    let border: Color
    let background: Color
    let text: Color
    let icon: Color
    let borderWidth: CGFloat

    static func resolve(
        isOpen: Bool,
        hasError: Bool,
        isEnabled: Bool,
        hasSelection: Bool
    ) -> IldsDropdownColors {
        if !isEnabled {
            return IldsDropdownColors(
                border: ILDSTokens.neutralCoolgray300,
                background: ILDSTokens.neutralCoolgray100,
                text: ILDSTokens.neutralCoolgray300,
                icon: ILDSTokens.neutralCoolgray300,
                borderWidth: 1
            )
        }
        if hasError {
            return IldsDropdownColors(
                border: ILDSTokens.errorRed600,
                background: ILDSTokens.globalWhite000,
                text: hasSelection ? ILDSTokens.neutralCoolgray900 : ILDSTokens.neutralCoolgray300,
                icon: ILDSTokens.errorRed600,
                borderWidth: isOpen ? 2 : 1
            )
        }
        if isOpen {
            return IldsDropdownColors(
                border: ILDSTokens.primaryOrange500,
                background: ILDSTokens.globalWhite000,
                text: hasSelection ? ILDSTokens.neutralCoolgray900 : ILDSTokens.neutralCoolgray300,
                icon: ILDSTokens.primaryOrange500,
                borderWidth: 2
            )
        }
        return IldsDropdownColors(
            border: ILDSTokens.neutralCoolgray200,
            background: ILDSTokens.globalWhite000,
            text: hasSelection ? ILDSTokens.neutralCoolgray900 : ILDSTokens.neutralCoolgray300,
            icon: ILDSTokens.neutralCoolgray400,
            borderWidth: 1
        )
    }
}
